### PR TITLE
Cross-Platform NamedLocks Reimplementation

### PR DESCRIPTION
**Pull request by tsavo-at-pieces**
Hey @bsutton - putting this here so that you and I can both start testing this together to make sure everything stays in tact. 

You likely know where the breaks would occur better than I but I wanted to share how I initially thought about refactoring dart_project.dart file with the updated NamedLocks guard. 

That said - if you're thinking it's really important for end users to still have access to the old NamedLocks class I can re-implement its' internals with the primitives that are found in [named_locks](https://github.com/open-runtime/named_locks) pretty easily I believe. The only caveat is that this new NamedLocks system runs on Native Semaphores across Unix and Windows i.e. building on the [native_semaphores](https://github.com/open-runtime/native_semaphores) codebase and doesn't rely on the filesystem i.e. there isn't support for file paths as Named Semaphores are OS wide. 

The other thing to note is that right now it supports cross process locking and reentrant behavior within a process. It is able to keep track of the counts reentrantly but it doesn't keep track of reentrant counts across process (yet). If this is needed by DCLI or the community this can be implemented with shared memory or mmap pretty easily. 

Let me know thoughts on some of this but most importantly I'm just looking to get your feedback on if it works as expected for the DCLI use case. Perhaps you can pull this branch down and do some testing in your environment and then circle back here. 

Let me know thoughts 🙏🤝

Cheers, 
-Tsavo

### CODE REVIEW

File: dcli/lib/src/script/dart_project.dart [hunk 1]
```diff
@@ -235,36 +228,43 @@ class DartProject {
...
-              extensionSearch: false,
-            );
-          } else {
-            // print(orange('Running pub get...'));
-            if (upgrade) {
-              await _pubupgrade();
+  void warmup({bool background = false, bool upgrade = false}) {
+    runtime.NamedLock.guard(
+      name: _lockName,
+      execution: runtime.ExecutionCall<void>(
```
[the remark this answers is not in the document]
Comment (tsavo-at-pieces) [reply dcli/lib/src/script/dart_project.dart:234]: And here is how it is leveraged within the guard function itself! 

https://github.com/open-runtime/named_locks/blob/3acd49b3fc0821e114abdd0e5461ccbb567faf89/lib/src/named_lock.dart#L157

File: dcli/lib/src/script/dart_project.dart [hunk 2]
```diff
@@ -235,36 +228,43 @@ class DartProject {
...
+              }
             }
+          } on PubGetException {
+            print(red("\ndcli warmup failed due to the 'pub get' call failing."));
           }
-        } on PubGetException {
-          print(red("\ndcli warmup failed due to the 'pub get' call failing."));
-        }
-      },
-      waiting: 'Waiting for warmup to complete...',
```
Comment (tsavo-at-pieces) [inline dcli/lib/src/script/dart_project.dart:266]: Can definitely add something like this to the NamedLock.guard input parameters if we wanted.

**Review by tsavo-at-pieces — COMMENTED** [review]
Let me know thoughts - just looking to shape this thing a bit more and have you take it through some of your local paces to see if it is behaving as expected. 

I have quite a lot of unit tests here for both NamedLock and NamedSemaphore its' underlying infrastructure. 

[NamedLock Tests](https://github.com/open-runtime/named_locks/blob/aot_monorepo_compat/test/named_lock_test.dart)

[NativeSemaphore Tests](https://github.com/open-runtime/native_semaphores/blob/aot_monorepo_compat/test/semaphore_reentrant_test.dart)

**Comment by bsutton** [comment]
So I would prefer to keep the existing NamedLock api as part of our public api so we don't unnecessarily break peoples code.

If a user wants to use the new guard method then my thinking is that we should direct them to use your package directly - this is also inline with dart recommendations.

**Comment by bsutton** [comment]
I'm also thinking that we need a new version of NamedLock - NamedLockAsync which allows the callback to be an async function.
This is in keeping with the likes of withTempDir/withTempDirAsync.

File: dcli/lib/src/script/dart_project.dart [hunk 3]
```diff
@@ -217,11 +210,11 @@ class DartProject {
     return null;
   }
 
-  NamedLock? __lock;
-  static const _lockName = 'script.lock';
+  // NamedLock? __lock;
+  static const _lockName = 'dcli.script.lock.a1';
```
[the remark this answers is not in the document]
Comment (bsutton) [reply dcli/lib/src/script/dart_project.dart:214]: so I assume this is just to reduce the chance of a conflict. If so it looks fine.

File: dcli/lib/src/script/dart_project.dart (hunk 2 above)
Comment (bsutton) [reply dcli/lib/src/script/dart_project.dart:266]: this comes back to my comments about retaining the existing NamedLocked API as part of our external API. This is nice little feature that makes it simple to provide user feedback.

File: dcli/lib/src/script/dart_project.dart [hunk 4]
```diff
@@ -321,8 +355,7 @@ class DartProject {
   ///
   void compile({bool install = false, bool overwrite = false}) {
```
[the remark this answers is not in the document]
Comment (bsutton) [reply dcli/lib/src/script/dart_project.dart:356]: maybe - which probably means that we should just protect it.  Dart probably has its own protections in place but given its easy for us to do I would be inclined to protect it.

File: dcli/lib/src/script/dart_project.dart (hunk 2 above)
Comment (tsavo-at-pieces) [reply dcli/lib/src/script/dart_project.dart:266]: Hey @bsutton working on this as we speak! No problem here - can absolutely keep the API in tact!

**Comment by bsutton** [comment]
> runs on Native Semaphores across Unix and Windows i.e. building on the [native_semaphores](https://github.com/open-runtime/native_semaphores) codebase and doesn't rely on the filesystem i.e. there isn't support for file paths as Named Semaphores are OS wide.

The files paths used for the locking are not part of the public api so the fact that you know use a different method is fine.

**Comment by bsutton** [comment]
> The other thing to note is that right now it supports cross process locking and reentrant behavior within a process. It is able to keep track of the counts reentrantly but it doesn't keep track of reentrant counts across process (yet). If this is needed by DCLI or the community this can be implemented with shared memory or mmap pretty easily.

The code should NOT be re-entrant across processes - so the current implementation is fine.

The code should ONLY be re-entrant within an isolate.  If two isolates in the same process try to enter a lock, only one is allowed to succeed. 

The way that re-entrance work is critical! 
If I've understood you correctly that multiple isolates in one process can enter the same lock, then we have a problem that will need to be resolved.

**Comment by bsutton** [comment]
So I'm a little confused - the PR title is linux only but then you state that the underlying implementation works on Windows and Linux?

I assume when you say linux you are including macos as well?

**Comment by tsavo-at-pieces** [comment]
> I'm also thinking that we need a new version of NamedLock - NamedLockAsync which allows the callback to be an async function. This is in keeping with the likes of withTempDir/withTempDirAsync.

Can absolutely do something like this - I actually allow the ExecutableCall callable to be sync or async and async values can be accessed on the completer.future  🙌 

[Here is the internal code](https://github.com/open-runtime/named_locks/blob/a3ccff60df95e35d185300e39419190bb5c30850/lib/src/execution_call.dart#L89) that does this, the completer field is higher up in that class and publicly accessible.

So to access it one could do something like: 

```
    final ExecutionCall<Future<bool>, Exception> _guarded = NamedLock.guard(
      name: name,
      execution: ExecutionCall<Future<bool>, Exception>(
        callable: () async {
          sleep(Duration(milliseconds: Random().nextInt(5000)));
          return Future.sync(true);
        },
      ),
    );

    print(await _guarded.completer.future);
    
    // or 
    print(await _guarded.returned);
```


Can absolutely do an NamedLock.guardAsync as well just to be super explicit!

**Comment by tsavo-at-pieces** [comment]
> > The other thing to note is that right now it supports cross process locking and reentrant behavior within a process. It is able to keep track of the counts reentrantly but it doesn't keep track of reentrant counts across process (yet). If this is needed by DCLI or the community this can be implemented with shared memory or mmap pretty easily.
> 
> The code should NOT be re-entrant across processes - so the current implementation is fine.
> 
> The code should ONLY be re-entrant within an isolate. If two isolates in the same process try to enter a lock, only one is allowed to succeed.
> 
> The way that re-entrance work is critical! If I've understood you correctly that multiple isolates in one process can enter the same lock, then we have a problem that will need to be resolved.

I believe the re-entrant behavior is as you described above. I have verified that all of the unit tests that you had previously are passing as expected 🤝 I will need to create some unit tests to double check that multiple isolates in one process **can not** enter the same lock but I am pretty sure this is the case as it stands right now!

**Comment by tsavo-at-pieces** [comment]
> So I'm a little confused - the PR title is linux only but then you state that the underlying implementation works on Windows and Linux?
> 
> I assume when you say linux you are including macos as well?

Yep - this weekend I'll remove that note on unix as Windows is 99% g2g I have one small nit in a unit test that I believe is a flipped boolean somewhere on the underlying named semaphore implementation. That said once that is passing the named locks code will work for windows with no changes needed 🫡

Screenshot 2024-05-11 at 6 14 54 PM

File: dcli/lib/src/script/dart_project.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (tsavo-at-pieces) [reply dcli/lib/src/script/dart_project.dart:356]: On it!

File: dcli/lib/src/script/dart_project.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (tsavo-at-pieces) [reply dcli/lib/src/script/dart_project.dart:214]: yes sir! I took the a1 off though as that was just a quick name change to run it on a different name

**Comment by bsutton** [comment]
I'm at the point that if you can get you code merged that we do a 4.0
release.

They are still a few dangling issue but I think we have the core paths that
most people use working.

As such I'm inclined to not let perfect get in the way of good enough.

Thoughts?

**Comment by tsavo-at-pieces** [comment]
> I'm at the point that if you can get you code merged that we do a 4.0
> release.
> 
> They are still a few dangling issue but I think we have the core paths that
> most people use working.
> 
> As such I'm inclined to not let perfect get in the way of good enough.
> 
> Thoughts?

Completely agree - hustling tonight to get it rockin and rollin! 

Shall we target a 4.0.0 release say EOW?

**Comment by tsavo-at-pieces** [comment]
@bsutton turns out that I had created unit tests to ensure multiple isolates in one process **can not** enter the same lock so we are good to go there 🙌

Link to unit test for that specifically found [here](https://github.com/open-runtime/native_semaphores/blob/599dc5b9b25d945b99260ffc1a0b57e4e4c37409/test/semaphore_reentrant_test.dart#L119)

**Comment by tsavo-at-pieces** [comment]
> So I'm a little confused - the PR title is linux only but then you state that the underlying implementation works on Windows and Linux?
> 
> I assume when you say linux you are including macos as well?

Windows support is good to go on the underlying native_semaphores package now 🎉

All tests passing [here](https://github.com/open-runtime/native_semaphores/actions/runs/9047751220) ✅

Screenshot 2024-05-11 at 9 19 23 PM

**Comment by bsutton** [comment]
Great, eow works for me too.

**Comment by tsavo-at-pieces** [comment]
> Great, eow works for me too.

Legendary 😎

Just verified that NamedLocks also is g2g on windows! 

Screenshot 2024-05-11 at 9 28 28 PM

Testing in DCLI now, PR on deck!

**Comment by bsutton** [comment]
So should I merge?

**Comment by tsavo-at-pieces** [comment]
Let me rip some testing and final adjustments and we should be g2g in next couple of hours MAX. Will report back in 30 with DCLI specific testing results!

**Comment by tsavo-at-pieces** [comment]
NamedLocks are looking solid - this may be good to merge here! 

See unit tests here:

[tool/run_specific_unit_tests.dart](https://github.com/tsavo-at-pieces/dcli/actions/workflows/run_specific_unit_tests.yml)

I updated the internals of DCLI NamedLock.withLock() and have a unit test preserved from before that used `withLock` and I also added one which used Runtime NamedLock.guard() directly as well 🙌

File: dcli/lib/src/script/dart_project.dart [hunk 5]
```diff
@@ -235,34 +223,37 @@ class DartProject {
...
+          } on PubGetException {
+            print(red("\ndcli warmup failed due to the 'pub get' call failing."));
           }
-        } on PubGetException {
-          print(red("\ndcli warmup failed due to the 'pub get' call failing."));
-        }
-      },
+        },
+      ),
       waiting: 'Waiting for warmup to complete...',
```
Comment (tsavo-at-pieces) [inline dcli/lib/src/script/dart_project.dart:257]: Added in the `waiting` property here!

File: dcli/lib/src/script/dart_project.dart [hunk 6]
```diff
@@ -356,21 +350,20 @@ class DartProject {
...
-      if (Shell.current.isSudo) {
-        /// bugger we just screwed the cache permissions so lets fix them.
-//         'chmod -R ${env['USER']}:${env['USER']} ${PubCache().pathTo}'.run;
-
-        throw DartProjectException(
-            'You must compile your script before running it under sudo');
-      }
-      pubUpgrade.run(compileExecutables: false);
-    });
+  void _pubupgrade() {
```
Comment (tsavo-at-pieces) [inline dcli/lib/src/script/dart_project.dart:362]: No longer async**

File: dcli/lib/src/script/dart_project.dart [hunk 7]
```diff
@@ -333,20 +328,19 @@ class DartProject {
...
-      final pubGet = PubGet(this);
-      if (Shell.current.isSudo) {
-        /// bugger we just screwed the cache permissions so lets fix them.
-//         'chmod -R ${env['USER']}:${env['USER']} ${PubCache().pathTo}'.run;
-        throw DartProjectException(
-            'You must compile your script before running it under sudo');
-      }
-      pubGet.run(compileExecutables: false);
-    });
+  void _pubget() {
```
Comment (tsavo-at-pieces) [inline dcli/lib/src/script/dart_project.dart:340]: No longer async**

File: dcli/lib/src/script/dart_project.dart [hunk 8]
```diff
@@ -235,34 +223,37 @@ class DartProject {
...
-                .start(
-              detached: true,
-              runInShell: true,
-              extensionSearch: false,
-            );
-          } else {
-            // print(orange('Running pub get...'));
-            if (upgrade) {
-              await _pubupgrade();
+  void warmup({bool background = false, bool upgrade = false}) {
```
Comment (tsavo-at-pieces) [inline dcli/lib/src/script/dart_project.dart:226]: No longer async**

File: dcli/lib/src/script/dart_project.dart [hunk 9]
```diff
@@ -276,36 +267,41 @@ class DartProject {
...
-
-        find('pubspec.lock', workingDirectory: pathToProjectRoot)
-            .forEach(delete);
-
-        find('*.dart', workingDirectory: pathToProjectRoot)
-            .forEach((scriptPath) {
-          final script = DartScript.fromFile(scriptPath);
-          if (exists(script.pathToExe)) {
-            delete(script.pathToExe);
+  void clean() {
```
Comment (tsavo-at-pieces) [inline dcli/lib/src/script/dart_project.dart:270]: No Longer Async**

File: dcli/lib/src/script/dart_project.dart [hunk 10]
```diff
@@ -320,10 +316,18 @@ class DartProject {
   /// [overwrite] defaults to false.
   ///
   void compile({bool install = false, bool overwrite = false}) {
-    find('*.dart', workingDirectory: pathToProjectRoot).forEach(
-      (file) => DartScript.fromFile(file)
-          .compile(install: install, overwrite: overwrite),
+    NamedLock.guard(
```
Comment (tsavo-at-pieces) [inline dcli/lib/src/script/dart_project.dart:319]: Added NamedLock.guard here just for good measure!

**Review by tsavo-at-pieces — COMMENTED** [review]
This very well may be ready for merging @bsutton 🤝

It seems that there are still some unit tests failing at large but at least for the NamedLocks stuff, I've triple checked everything here and the entire supply chain has super robust unit tests as well. 

Really appreciate the patience and I think this will be rocking and rollin in the long term 🦾

Also - we probably got a performance boost moving away from UDP/TCP Sockets and File reads & writes!

I have a few more features coming down the line for the NamedLock package but as it stands I am feeling pretty darn confident about it. 

Let me know thoughts & cheers brotha 🍻 

Myself and the community appreciate all you do!

**Comment by bsutton** [comment]
I'm going to target Wednesday for a release as I will have some spare time. 

Will try to clean up a few more if the failing unit tests. 
These are mostly problems with the tests.

**Pull request merged**